File: stomp/StOMP.m
% Stagewise Orthogonal Matching Pursuit
% This function returns a column vector recovered using StOMP algorithm.
% length(x)=size(A,2);
% A--> measurement matrix
% y--> test vector: y=Ax
% N--> Number of iterations/stages
function x = StOMP(A,y,N)
    r=y; % initial Residue
    O=[]; % initialisation of Support Vectors
    n=size(A,2);
    x=zeros(n,1); % initialisation of x
    t=3; % threshold parameter
    sd=norm(r);
    for i=1:N
        if A*x==y
            break
        end
        c=A'*r; % Correlation
        sd=norm(r)/sqrt(n); % noise level(standard deviation)
        ind=find(abs(c)>=t*sd); % find the desired indices greater than threshold
        O=union(O,ind); % Update Support
        Ao=A(:,O); % Updated measurement matrix
        x1=Ao\y; % min ||y-Ao*x||
        r=y-Ao*x1; % ith step residual
        x(O)=x1;
    end
end

File: stomp.py
"""
Stagewise Orthogonal Match Pursuit Sparse Signal Recovery

References:
    D. Donoho, Y. Tsaig, I. Drori, J Starck, "Sparse Solution of Underdetermined Linear Equations
    by Stagewise Orthogonal Matching Pursuit,", March 2006

    Hammeed, Maxin Abdulrasool, "Comparative Analysis of Orthogonal Matching Pursuit
    and least angle regression," Michigan State University, A Thesis for Masters of
    Science, 2012.

    Angshul Majumdar, "Compressed Sensing for Engineers", CRC Press, 3 Dec 2018.
"""

import numpy as np
import scipy.linalg as lin

def stomp(y, A, sigma, N):
    """
    stagewise orthogonal match pursuit with configurable termination criteria

    Parameters:
        y: `compressed samples`
        A: `sampling matrix`
        sigma: `number of std deviations (2-3)`
        N: `number of iterations`

    Returns:
        x_hat: `reconstructed signal`
    """

    r = np.copy(y)       # init residual
    lamda = set()        # set of support indicies
    x = np.array([])     # null output init
    n = A.shape[1]       # size of solution

    for t in range(N):
        c = np.dot(A.T, r)                  # correlation
        thr = sigma * np.std(r)             # compute number of std devs for threshold
        inds = np.nonzero(np.abs(c) > thr)  # find abs support above threshold
        lamda.update(inds[0].tolist())      # update support

        P = A[:, list(lamda)]                     # compose the submatrix
        x = np.linalg.lstsq(P, y, rcond=None)[0]  # min ||y-P*x||_2
        r = y - np.dot(P, x)                      # update residual

    # embed coefficients in support
    x_hat = np.zeros((n, 1))
    x_hat[list(lamda)] = x

    return x_hat
